GVFS/GVFS.Common/Cleanup/GitCleanupScheduler.cs

```diff
@@ -0,0 +1,42 @@
+﻿using GVFS.Common.Git;
+using System;
+using System.Threading;
+
+namespace GVFS.Common.Cleanup
+{
+    public class GitCleanupScheduler : IDisposable
+    {
+        private readonly TimeSpan prefetchPeriod = TimeSpan.FromMinutes(15);
+        private Timer prefetchStepTimer;
+        private GVFSContext context;
+        private GitObjects gitObjects;
+        private GitCleanupQueue queue;
+
+        public GitCleanupScheduler(GVFSContext context, GitObjects gitObjects, GitCleanupQueue queue)
+        {
+            this.context = context;
+            this.gitObjects = gitObjects;
+            this.queue = queue;
+
+            this.ScheduleRecurringCleanupSteps();
+        }
+
+        public void Dispose()
+        {
+            this.prefetchStepTimer?.Dispose();
+            this.prefetchStepTimer = null;
+        }
+
+        public void ScheduleRecurringCleanupSteps()
+        {
+            if (!this.context.Unattended && this.gitObjects.IsUsingCacheServer())
+            {
+                this.prefetchStepTimer = new Timer(
+                (state) => this.queue.Enqueue(new PrefetchStep(this.context, this.gitObjects, requireCacheLock: true)),

```

Comment:
I realize we only have one of these right now but since the method name has `Steps` should we have a list of timers and recurring steps so it is easier to add to this later on.  I'm okay with waiting until we do have more than one to make this change though.  

The reason I kind of thought about this was because I was wondering why it was newing up the PrefetchStep every time the timer goes off.  Could this be a member variable?

Reply:
When we have a list of three or four recurring steps, where do you want those specific steps to be added to the schedule? Do you want the `InProcessMount` to be responsible for that, or a third class that initializes the scheduler?

To me, pulling out the choices of _what_ is scheduled away from the logic of _how_ it recurs is unnecessary complication.

Reply:
I started playing with this, but we actually want to be very careful with the schedule here.

* We don't want any background steps when unattended.

* We don't want to run prefetch automatically if we don't have a cache server. We'll still want to run the other steps, though.

* The other steps will want a shorter `dueTime` than their `period`, as we want them to run at most once per day, but want the steps to run when people restart their machine nightly.

I can make this class slightly more generic by creating a `List<Timer>` of timers, and add the timers to the list as necessary. Those timers can then be disposed from the list instead of one at a time.